sql LOOKUP() failure when filtering on a key value that is not present in the lookup

LOOKUP() bombs when you add a WHERE on the key, and you specify something that's not in the lookup table.

I have a small lookup, "myLookup", with this map:

```
{
  "#ar.wikipedia": "wikipedia lookup value"
}
```

and loaded wikipedia.  This query works (the channel in the WHERE exists in the lookup):

```
SELECT channel, LOOKUP(channel, 'myLookup')
FROM wikipedia
WHERE channel = '#ar.wikipedia'
```

This one gives an error (the channel in the WHERE does not exist in the lookup):

```
SELECT channel, LOOKUP(channel, 'myLookup')
FROM wikipedia
WHERE channel = 'test'
```

The error is:

```
javax.servlet.ServletException: java.lang.AssertionError: Cannot add expression of different type to set: set type is RecordType(VARCHAR channel, VARCHAR NOT NULL EXPR$1) NOT NULL expression type is RecordType(VARCHAR channel, VARCHAR EXPR$1) NOT NULL set is rel#6110:LogicalProject.NONE.[](input=HepRelVertex#6109,channel=$0,EXPR$1=LOOKUP($0, &apos;myLookup&apos;)) expression is LogicalProject(channel=[CAST(&apos;test&apos;:VARCHAR):VARCHAR], EXPR$1=[null:VARCHAR]) LogicalFilter(condition=[=($0, &apos;test&apos;)]) LogicalProject(channel=[$1]) LogicalTableScan(table=[[druid, wikipedia]])
```

The same logic will work as a JOIN, and properly return no results:

```
SELECT channel, v
FROM wikipedia
JOIN lookup.myLookup
ON channel = k
WHERE channel = 'test'
```

--- a/sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java
+++ b/sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java
@@ -40,7 +40,7 @@ public class QueryLookupOperatorConversion implements SqlOperatorConversion
   private static final SqlFunction SQL_FUNCTION = OperatorConversions
       .operatorBuilder("LOOKUP")
       .operandTypes(SqlTypeFamily.CHARACTER, SqlTypeFamily.CHARACTER)
-      .returnTypeNonNull(SqlTypeName.VARCHAR)
+      .returnTypeNullable(SqlTypeName.VARCHAR)
       .functionCategory(SqlFunctionCategory.STRING)
       .build();
 

--- a/sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java
+++ b/sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java
@@ -17091,26 +17091,36 @@ public class CalciteQueryTest extends BaseCalciteQueryTest
   @Test
   public void testLookupWithNull() throws Exception
   {
+    List<Object[]> expected;
+    if (useDefault) {
+      expected = ImmutableList.<Object[]>builder().add(
+          new Object[]{NULL_STRING, NULL_STRING},
+          new Object[]{NULL_STRING, NULL_STRING},
+          new Object[]{NULL_STRING, NULL_STRING}
+      ).build();
+    } else {
+      expected = ImmutableList.<Object[]>builder().add(
+          new Object[]{NULL_STRING, NULL_STRING},
+          new Object[]{NULL_STRING, NULL_STRING}
+      ).build();
+    }
     testQuery(
-        "SELECT dim1 ,lookup(dim1,'lookyloo') from foo where dim1 = 'def'",
+        "SELECT dim2 ,lookup(dim2,'lookyloo') from foo where dim2 is null",
         ImmutableList.of(
             new Druids.ScanQueryBuilder()
-            .dataSource(CalciteTests.DATASOURCE1)
-            .intervals(querySegmentSpec(Filtration.eternity()))
-            .virtualColumns(
-                expressionVirtualColumn("v0", "'def'", ValueType.STRING),
-                expressionVirtualColumn("v1", "null", ValueType.STRING)
-            )
-            .columns("v0", "v1")
-            .legacy(false)
-            .filters(new SelectorDimFilter("dim1", "def", null))
-            .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
-            .context(QUERY_CONTEXT_DEFAULT)
-            .build()
+                .dataSource(CalciteTests.DATASOURCE1)
+                .intervals(querySegmentSpec(Filtration.eternity()))
+                .virtualColumns(
+                    expressionVirtualColumn("v0", "null", ValueType.STRING)
+                )
+                .columns("v0")
+                .legacy(false)
+                .filters(new SelectorDimFilter("dim2", NULL_STRING, null))
+                .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
+                .context(QUERY_CONTEXT_DEFAULT)
+                .build()
         ),
-        ImmutableList.<Object[]>builder().add(
-            new Object[]{"def", NULL_STRING}
-        ).build()
+        expected
     );
   }
 }

--- a/sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java
+++ b/sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java
@@ -17092,25 +17092,24 @@ public class CalciteQueryTest extends BaseCalciteQueryTest
   public void testLookupWithNull() throws Exception
   {
     testQuery(
-        "SELECT dim2 ,lookup(dim2,'lookyloo') from foo where dim2 is null",
+        "SELECT dim1 ,lookup(dim1,'lookyloo') from foo where dim1 = 'def'",
         ImmutableList.of(
             new Druids.ScanQueryBuilder()
             .dataSource(CalciteTests.DATASOURCE1)
             .intervals(querySegmentSpec(Filtration.eternity()))
             .virtualColumns(
-                expressionVirtualColumn("v0", "null", ValueType.STRING)
+                expressionVirtualColumn("v0", "'def'", ValueType.STRING),
+                expressionVirtualColumn("v1", "null", ValueType.STRING)
             )
-            .columns("v0")
+            .columns("v0", "v1")
             .legacy(false)
-            .filters(new SelectorDimFilter("dim2", "", null))
+            .filters(new SelectorDimFilter("dim1", "def", null))
             .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
             .context(QUERY_CONTEXT_DEFAULT)
             .build()
         ),
         ImmutableList.<Object[]>builder().add(
-            new Object[]{"", NULL_STRING},
-            new Object[]{"", NULL_STRING},
-            new Object[]{"", NULL_STRING}
+            new Object[]{"def", NULL_STRING}
         ).build()
     );
   }

--- a/sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java
+++ b/sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java
@@ -17086,4 +17086,32 @@ public class CalciteQueryTest extends BaseCalciteQueryTest
         ).build()
     );
   }
+
+
+  @Test
+  public void testLookupWithNull() throws Exception
+  {
+    testQuery(
+        "SELECT dim2 ,lookup(dim2,'lookyloo') from foo where dim2 is null",
+        ImmutableList.of(
+            new Druids.ScanQueryBuilder()
+            .dataSource(CalciteTests.DATASOURCE1)
+            .intervals(querySegmentSpec(Filtration.eternity()))
+            .virtualColumns(
+                expressionVirtualColumn("v0", "null", ValueType.STRING)
+            )
+            .columns("v0")
+            .legacy(false)
+            .filters(new SelectorDimFilter("dim2", "", null))
+            .resultFormat(ScanQuery.ResultFormat.RESULT_FORMAT_COMPACTED_LIST)
+            .context(QUERY_CONTEXT_DEFAULT)
+            .build()
+        ),
+        ImmutableList.<Object[]>builder().add(
+            new Object[]{"", NULL_STRING},
+            new Object[]{"", NULL_STRING},
+            new Object[]{"", NULL_STRING}
+        ).build()
+    );
+  }
 }

